### PR TITLE
A few push fixes

### PR DESCRIPTION
**Pull request by adammhaile**
@tbielawa @sosiouxme 

- Added back push retry
- Added stateful logic to push
- not switches branch on distgit clone when rebase skipped

### CODE REVIEW

File: doozer [hunk 1]
```diff
@@ -942,12 +942,44 @@ def images_push(runtime, tag, version_release, to_defaults, late_only, to, dry_r
...
     version_release_tuple = None
 
     if version_release:
         version_release_tuple = version_release.split('-')
         click.echo('Setting up to push: version={} release={}'.format(version_release_tuple[0], version_release_tuple[1]))
 
+    items = runtime.image_metas()
+
+    if required:
+        msg = 'The following images failed during the previous step and are required:\n{}'.format('\n'.join(required))
```
Comment (tbielawa) [inline doozer]: Same as `lstate['msg']`?

**Comment by tbielawa** [comment]
/lgtm
Just curious about that one thing

File: doozer [hunk 2]
```diff
@@ -942,12 +942,44 @@ def images_push(runtime, tag, version_release, to_defaults, late_only, to, dry_r
...
+    required = []
+    pre_failed = []
+
+    if pre_step:
+        for img, status in pre_step['images'].iteritems():
+            if isinstance(status, str):
+                img_obj = runtime.image_map[img]
+                pre_failed.append(img)
+                if img_obj.required:
+                    required.append[img]
```
[the remark this answers is not in the document]
Comment (tbielawa) [reply doozer]: nice catch

File: doozer [hunk 3]
```diff
@@ -942,12 +942,44 @@ def images_push(runtime, tag, version_release, to_defaults, late_only, to, dry_r
     runtime.state[cmd] = dict(state.TEMPLATE_IMAGE)
     lstate = runtime.state[cmd]  # get local convenience copy
 
+    pre_step = None
```
[the remark this answers is not in the document]
Comment (adammhaile) [reply doozer]: Yeah yeah... I know :P Fixed.

File: doozer (hunk 2 above)
[the remark this answers is not in the document]
Comment (tbielawa) [reply doozer]: This is why we need basic pyflakes on these repos

[the remark this answers is not in the document]
Comment (adammhaile) [reply doozer]: Hmm... that was copy/pasted from elsewhere. You found a bug :)

[the remark this answers is not in the document]
Comment (adammhaile) [reply doozer]: @tbielawa I run flake on it (well, my editor does). It wouldn't have caught that, purely a runtime error. It *could* be valid. Fixed in both places now.

File: doozer (hunk 1 above)
Comment (adammhaile) [reply doozer]: Hmm... yeah, not sure why I did that. Fixing...

**Pull request merged**